core/src/cli_commands/update_pool.rs

```diff
@@ -0,0 +1,629 @@
+        &new_pool_contract
+            .ergo_tree()
+            .sigma_serialize_bytes()
+            .unwrap(),
+    );
+    println!("Pool Box Hash (new): {}", String::from(pool_box_hash));
+    println!(
+        "Reward Token ID (old): {}",
+        String::from(old_oracle_config.token_ids.reward_token_id.clone())
+    );
+    println!(
+        "Reward Token ID (new): {}",
+        String::from(new_oracle_config.token_ids.reward_token_id.clone())
+    );
+    println!(
+        "Reward Token Amount (old): {}",
+        old_pool_box.reward_token().amount.as_u64()
+    );
+    println!("Reward Token Amount (new): {}", new_tokens.amount.as_u64());
+    println!(
+        "Update NFT ID (old): {}",
+        String::from(old_pool_box.contract().update_nft_token_id().clone())
+    );
+    println!(
+        "Update NFT ID (new): {}",
+        String::from(new_pool_contract.update_nft_token_id().clone())
+    );
+    println!(
+        "Refresh NFT ID (old): {}",
+        String::from(old_pool_box.contract().refresh_nft_token_id().clone())
+    );
+    println!(
+        "Refresh NFT ID (new): {}",
+        String::from(new_pool_contract.refresh_nft_token_id().clone())
+    );
+}
+
+#[allow(clippy::too_many_arguments)]
+fn build_update_pool_box_tx(
+    pool_box_source: &dyn PoolBoxSource,
+    ballot_boxes: &dyn BallotBoxesSource,
+    wallet: &dyn WalletDataSource,
+    update_box: &dyn UpdateBoxSource,
+    new_pool_contract: PoolContract,
+    new_reward_tokens: Option<Token>,
+    height: u32,
+    change_address: Address,
+) -> Result<TransactionContext<UnsignedTransaction>, UpdatePoolError> {
+    let min_votes = update_box.get_update_box()?.min_votes();
+    let old_pool_box = pool_box_source.get_pool_box()?;
+    let update_box = update_box.get_update_box()?;
+    let pool_box_hash = Constant::from(blake2b256_hash(
+        &new_pool_contract
+            .ergo_tree()
+            .sigma_serialize_bytes()
+            .unwrap(),
+    ));
+    let reward_tokens = new_reward_tokens.unwrap_or_else(|| old_pool_box.reward_token());
+    // Find ballot boxes that are voting for the new pool hash
+    let mut sorted_ballot_boxes = ballot_boxes.get_ballot_boxes()?;
+    // Sort in descending order of ballot token amounts. If two boxes have the same amount of ballot tokens, also compare box value, in case some boxes were incorrectly created below minStorageRent
+    sorted_ballot_boxes.sort_by(|b1, b2| {
+        (
+            *b1.ballot_token().amount.as_u64(),
+            *b1.get_box().value.as_u64(),
+        )
+            .cmp(&(
+                *b2.ballot_token().amount.as_u64(),
+                *b2.get_box().value.as_u64(),
+            ))
+    });
+    sorted_ballot_boxes.reverse();
+
+    let mut votes_cast = 0;
+    let vote_ballot_boxes: Vec<BallotBoxWrapper> = ballot_boxes
+        .get_ballot_boxes()?
+        .into_iter()
+        .filter(|ballot_box| {
+            let ballot_box = ballot_box.get_box();
+            ballot_box
```

Comment:
Hmm, one way would be to change BallotBoxesSource to VoteBallotBoxesSource, with the following parameters

```rust
pub trait VoteBallotBoxesSource {
    // Return ballot boxes that have voted for the following parameters
    fn get_vote_ballot_boxes(&self, pool_box_hash: Digest32, reward_tokens: Option<Token>, update_box_height: Option<u32>)  -> Result<Vec<BallotBox>>
}
```
I'm not opposed to having a seperate type for VoteBallotBox though

Reply:
Well, in this case, it sounds like it can query ballot boxes by voting parameters which is not our case. We want to source to give us the "current" (unspent) voting ballot boxes. The source can provide them without asking for any parameters.
I meant to make voting and voted boxes different types so that we don't need to check if `BallotBox` is voting or not in the update pool command. So the update pool command gets only voting ballot boxes, whereas the "vote update pool" command can get both.